nano/node/transport/tcp_server.cpp

```diff
@@ -10,59 +10,57 @@
 
 #include <memory>
 
+namespace
+{
+bool is_temporary_error (boost::system::error_code const & ec_a)
+{
+	switch (ec_a.value ())
+	{
+#if EAGAIN != EWOULDBLOCK
+		case EAGAIN:
+#endif
+
+		case EWOULDBLOCK:
+		case EINTR:
+			return true;
+		default:
+			return false;
+	}
+}
+}
+
 /*
  * tcp_listener
  */
 
-nano::transport::tcp_listener::tcp_listener (uint16_t port_a, nano::node & node_a) :
+nano::transport::tcp_listener::tcp_listener (uint16_t port_a, nano::node & node_a, std::size_t max_inbound_connections) :
 	node (node_a),
-	port (port_a)
+	strand{ node_a.io_ctx.get_executor () },
+	acceptor{ node_a.io_ctx },
+	local{ boost::asio::ip::tcp::endpoint{ boost::asio::ip::address_v6::any (), port_a } },
+	max_inbound_connections{ max_inbound_connections }
 {
 }
 
-void nano::transport::tcp_listener::start ()
+void nano::transport::tcp_listener::start (std::function<bool (std::shared_ptr<nano::transport::socket> const &, boost::system::error_code const &)> callback_a)
 {
 	nano::lock_guard<nano::mutex> lock{ mutex };
 	on = true;
-	listening_socket = std::make_shared<nano::transport::server_socket> (node, boost::asio::ip::tcp::endpoint (boost::asio::ip::address_v6::any (), port), node.config.tcp_incoming_connections_max);
+	acceptor.open (local.protocol ());
+	acceptor.set_option (boost::asio::ip::tcp::acceptor::reuse_address (true));
 	boost::system::error_code ec;
-	listening_socket->start (ec);
-	if (ec)
+	acceptor.bind (local, ec);
+	if (!ec)
 	{
-		node.logger.always_log (boost::str (boost::format ("Network: Error while binding for incoming TCP/bootstrap on port %1%: %2%") % listening_socket->listening_port () % ec.message ()));
-		throw std::runtime_error (ec.message ());
+		acceptor.listen (boost::asio::socket_base::max_listen_connections, ec);
 	}
-
-	// the user can either specify a port value in the config or it can leave the choice up to the OS:
-	// (1): port specified
-	// (2): port not specified
-	//
-	const auto listening_port = listening_socket->listening_port ();
```

Comment:
This block of code was written by Theo and deals with setting the port number in a semi automatic way. I remember he spent a lot of time on this. What are we losing by removing this? We are probably losing the ability to automatically let the operating system give a node a listening address. I do not know what depends on that ability though and whether it is important.

Reply:
The ability to have an OS selected port remains. This is done by setting the peering_port=0 or on the command line --config node.peering_port=0.

this->local is set in the constructor to the port passed in to the constructor and just above you can see the acceptor binds to the specified port.

Tests make use of the OS-selected port so tests would be failing but I also verified by running a node and overriding the peering_port to 0 and checking the port it's bound to:

nano_node 69415      clemahieu   18u  IPv6 0x5bc40e581518c7f9      0t0    TCP *:54688 (LISTEN)

The previous functionality was copying the port selected in to this->port and using it only in tcp_listener::endpoint (). endpoint() was changed to directly query the listening port from the acceptor rather than caching a copy locally.